lib/Dialect/TT/IR/TTOpsTypes.cpp

```diff
@@ -881,6 +897,36 @@ DeviceAttr DeviceAttr::get(::mlir::MLIRContext *context,
   return get(context, systemDesc, chipIds);
 }
 
+// Sample the last index in the tensor to get the last addressable element of
+// the tensor.
+uint64_t DeviceAttr::getLayoutSizeBytes(ArrayRef<int64_t> scalarShape,
+                                        LayoutAttr layout,
+                                        MemorySpace memorySpace) const {
+  SmallVector<int64_t> shape = layout.isTiled()
+                                   ? layout.getTiledShape(scalarShape)
+                                   : SmallVector<int64_t>(scalarShape);
+  AffineMap linearMap =
+      layout.isTiled() ? layout.getIdentityTileLinearMap() : layout.getLinear();
+  mlir::SmallVector<std::int64_t> linearShape =
+      ttmlir::utils::evalShape(linearMap, shape);
+  AffineMap memoryMap = layout.replaceMemoryMapSymbolsWithShardShape(
+      getMapForMemorySpace(memorySpace));
+  mlir::SmallVector<std::int64_t> physicalMemory =
+      ttmlir::utils::evalShape(memoryMap, linearShape);
+  std::int64_t elementSize = layout.getElementSizeBytes();
+  uint64_t sizeBytes =
+      physicalMemory[MemoryMapResultIdx::ShardOffset] * elementSize;
+  return sizeBytes;
+}
+
+uint64_t DeviceAttr::getTensorSizeBytes(RankedTensorType tensorType,
```

Comment:
Is this returning memory size of full Tensor, its not per shard(unit of allocation)?

Reply:
It's a bit nuanced, I wasn't sure how to name the function because it depends on the memory space:
- L1: It's exactly the shard size
- DRAM: Is more nuanced because the shards are arranged row major into, for e.g. a linear 12GB range.  But then that whole blob gets paged and interleaved between all dram channels, due to paging and rounding the footprint ends up being close to "the whole tensor" / 12 (for 12 dram channels).
- System: Unsupported, but theoretically this would be the whole tensor regardless of shard size.  We might want to allocate from this memory space in metal direct + huge page.

Reply:
Awesome, it's great you checked-in this as a comment, there are a few more util functions which could benefit from this. And now based on your comment I figured this is exactly what I need.